synapse/storage/schema/main/delta/61/01insertion_event_lookups.sql

```diff
@@ -0,0 +1,50 @@
+/* Copyright 2021 The Matrix.org Foundation C.I.C
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- Add a table that keeps track of "insertion" events back in the history
+-- when we get a "marker" event over the "live" timeline. When navigating the DAG
+-- and we hit an event which matches `insertion_prev_event_id`, it should backfill 
+-- the "insertion" event and start navigating from there.
+
+CREATE TABLE IF NOT EXISTS insertion_events(
+    insertion_event_id TEXT NOT NULL,
+    room_id TEXT NOT NULL,
+    next_chunk_id TEXT NOT NULL,
+    UNIQUE (insertion_event_id, room_id, next_chunk_id)
+);
+
+CREATE INDEX IF NOT EXISTS insertion_events_insertion_room_id ON insertion_events(room_id);
+CREATE INDEX IF NOT EXISTS insertion_events_insertion_event_id ON insertion_events(insertion_event_id);
+CREATE INDEX IF NOT EXISTS insertion_events_next_chunk_id ON insertion_events(next_chunk_id);
+
+CREATE TABLE IF NOT EXISTS insertion_event_edges(
+    insertion_event_id TEXT NOT NULL,
+    room_id TEXT NOT NULL,
+    insertion_prev_event_id TEXT NOT NULL,
+    UNIQUE (insertion_event_id, room_id, insertion_prev_event_id)
+);
+
+CREATE INDEX IF NOT EXISTS insertion_event_edges_insertion_room_id ON insertion_event_edges(room_id);
+CREATE INDEX IF NOT EXISTS insertion_event_edges_insertion_event_id ON insertion_event_edges(insertion_event_id);
+CREATE INDEX IF NOT EXISTS insertion_event_edges_insertion_prev_event_id ON insertion_event_edges(insertion_prev_event_id);
+
+CREATE TABLE IF NOT EXISTS chunk_edges(
+    event_id TEXT NOT NULL,
+    room_id TEXT NOT NULL,
+    chunk_id TEXT NOT NULL,
+    UNIQUE (event_id, room_id)
+);
+
+CREATE INDEX IF NOT EXISTS chunk_edges_chunk_id ON chunk_edges(chunk_id);
```

Comment:
I saw the following error once when running the Complement tests and just took note of it. Not quite sure if flakey or random intermittent state of my code while I was debugging stuff.

```
2021-07-13 02:27:57,810 - synapse.handlers.federation - 1248 - ERROR - GET-4 - Failed to backfill from hs1 because NOT NULL constraint failed: event_to_state_groups.state_group
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/federation.py", line 1216, in try_backfill
    await self.backfill(
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/federation.py", line 1035, in backfill
    await self._auth_and_persist_event(dest, event, context, backfilled=True)
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/federation.py", line 2222, in _auth_and_persist_event
    await self._run_push_actions_and_persist_event(event, context, backfilled)
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/federation.py", line 2244, in _run_push_actions_and_persist_event
    await self.persist_events_and_notify(
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/federation.py", line 3290, in persist_events_and_notify
    events, max_stream_token = await self.storage.persistence.persist_events(
  File "/usr/local/lib/python3.8/site-packages/synapse/logging/opentracing.py", line 774, in _trace_inner
    return await func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/persist_events.py", line 320, in persist_events
    ret_vals = await yieldable_gather_results(enqueue, partitioned.items())
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/persist_events.py", line 237, in handle_queue_loop
    ret = await self._per_item_callback(
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/persist_events.py", line 577, in _persist_event_batch
    await self.persist_events_store._persist_events_and_state_updates(
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/databases/main/events.py", line 176, in _persist_events_and_state_updates
    await self.db_pool.runInteraction(
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/database.py", line 681, in runInteraction
    result = await self.runWithConnection(
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/database.py", line 770, in runWithConnection
    return await make_deferred_yieldable(
  File "/usr/local/lib/python3.8/site-packages/twisted/python/threadpool.py", line 238, in inContext
    result = inContext.theWork()  # type: ignore[attr-defined]
  File "/usr/local/lib/python3.8/site-packages/twisted/python/threadpool.py", line 254, in <lambda>
    inContext.theWork = lambda: context.call(  # type: ignore[attr-defined]
  File "/usr/local/lib/python3.8/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/local/lib/python3.8/site-packages/twisted/python/context.py", line 83, in callWithContext
    return func(*args, **kw)
  File "/usr/local/lib/python3.8/site-packages/twisted/enterprise/adbapi.py", line 293, in _runWithConnection
    compat.reraise(excValue, excTraceback)
  File "/usr/local/lib/python3.8/site-packages/twisted/python/deprecate.py", line 298, in deprecatedFunction
    return function(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/twisted/python/compat.py", line 403, in reraise
    raise exception.with_traceback(traceback)
  File "/usr/local/lib/python3.8/site-packages/twisted/enterprise/adbapi.py", line 284, in _runWithConnection
    result = func(conn, *args, **kw)
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/database.py", line 765, in inner_func
    return func(db_conn, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/database.py", line 549, in new_transaction
    r = func(cursor, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/synapse/logging/utils.py", line 69, in wrapped
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/databases/main/events.py", line 385, in _persist_events_txn
    self._store_event_state_mappings_txn(txn, events_and_contexts)
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/databases/main/events.py", line 2065, in _store_event_state_mappings_txn
    self.db_pool.simple_insert_many_txn(
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/database.py", line 923, in simple_insert_many_txn
    txn.execute_batch(sql, vals)
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/database.py", line 280, in execute_batch
    self.executemany(sql, args)
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/database.py", line 300, in executemany
    self._do_execute(self.txn.executemany, sql, *args)
  File "/usr/local/lib/python3.8/site-packages/synapse/storage/database.py", line 330, in _do_execute
    return func(sql, *args)
sqlite3.IntegrityError: NOT NULL constraint failed: event_to_state_groups.state_group
```

Reply:
This probably has something to do with how we are pulling in the historical messages and the auth events not being there 🤔

Reply:
This is definitely a problem but it's flakey and doesn't happen every time which makes it harder to debug.

It looks like it's failing for one of the historical messages occasionally:

Logged from [`synapse/storage/databases/main/events.py#L1942-L1962`](https://github.com/matrix-org/synapse/blob/0ae95b38474a4d64a4d5057499e645a3b81e3736/synapse/storage/databases/main/events.py#L1942-L1962)
```
state_groups dict_items([('$SKA-gTylAHOHjmy5yScvHvPcTr7FikhGxVMmCVR6ufw', None)])
```

And the full event for context:
```
{
  "type": "m.room.message",
  "room_id": "!WLfsCaykamkOnZMHxt:hs1",
  "sender": "@maria:hs1",
  "content": {
    "body": "Historical 1 (chunk=0)",
    "msgtype": "m.text",
    "org.matrix.msc2716.historical": true,
    "org.matrix.msc2716.chunk_id": "ndocOxvR"
  },
  "origin_server_ts": 1626324541183,
  "unsigned": {
    "age": 1360
  },
  "event_id": "$SKA-gTylAHOHjmy5yScvHvPcTr7FikhGxVMmCVR6ufw",
  "user_id": "@maria:hs1",
  "age": 1360
},
```

---

I can see that hs2 does a `hs1:event_auth/` check for that event but then later says `Skipping auth_event fetch for outlier` which is expected. Then my guess is somehow we are getting to the `state_groups` code which doesn't like `outliers` with no state 🤷‍♀️

```
2021-07-15 04:49:03,643 - synapse.http.matrixfederationclient - 278 - INFO - GET-7 - {GET-O-17} [hs1] Completed request: 200 OK in 0.01 secs, got 3359 bytes - GET matrix://hs1/_matrix/federation/v1/event_auth/%21WLfsCaykamkOnZMHxt%3Ahs1/%24SKA-gTylAHOHjmy5yScvHvPcTr7FikhGxVMmCVR6ufw
2021-07-15 04:49:03,648 - synapse.handlers.federation - 2733 - INFO - GET-7 - Skipping auth_event fetch for outlier
2021-07-15 04:49:03,650 - synapse.storage.databases.main.events - 2030 - INFO - persist_events-27 - state_groups asdfasdf dict_items([])
2021-07-15 04:49:03,653 - synapse.storage.databases.main.events - 2030 - INFO - persist_events-27 - state_groups asdfasdf dict_items([('$SKA-gTylAHOHjmy5yScvHvPcTr7FikhGxVMmCVR6ufw', None)])
2021-07-15 04:49:03,655 - synapse.handlers.federation - 1244 - ERROR - GET-7 - Failed to backfill from hs1 because NOT NULL constraint failed: event_to_state_groups.state_group
```

Reply:
@erikjohnston Does every `event` need a `state_group`? I'm curious if I can just add a condition that if the `event` doesn't have a `state_group`, then just skip trying to [persist it to `event_to_state_groups`](https://github.com/matrix-org/synapse/blob/0ae95b38474a4d64a4d5057499e645a3b81e3736/synapse/storage/databases/main/events.py#L1939).

Reply:
It looks like a `state_group` isn't necessary for outliers because we don't have the state for them. Our historical state to auth are outliers but the actual historical messages are just backfilled events.

> _state_group:
> [... ]
> For an outlier, where we don't have the state at the event, this will be None.
>
> [`synapse/events/snapshot.py#L45`](https://github.com/matrix-org/synapse/blob/323452944e13114264b0c645db875dc6950315c5/synapse/events/snapshot.py#L45)

Reply:
We should really document this somewhere, but: for every non-outlier event we need to know the state at that event. Instead of storing the full state for each event in the DB (i.e. a `event_id -> state` mapping), which is *very* space inefficient when state doesn't change, we instead assign each different set of state a "state group" and then have mappings of `event_id -> state_group` and `state_group -> state`

Reply:
Thanks for the clarification @erikjohnston, I've fixed this up in https://github.com/matrix-org/synapse/pull/10439

Reply:
Documented in https://github.com/matrix-org/synapse/pull/10464 🎉